### ISSUE TITLE
Error in the filterChannels() function

### ISSUE THREAD

**Issue by cmcg13** [wearefine/slack-archive-bot#7]
*Edited*

Hey folks,

I was debugging an issue where running this bot archived all channels regardless of the date of the latest comment. I can now see the error:

On Line 103 of Index.js - the array that has been passed through the getHistory method is being combined with the original get all slack.channels.list array. So it's just adding in everything which was originally obtained and passing them all through to be archived (minus those passed through to neverArchive array).

Once I corrected this concat line - the bot works as I would expect.

**Comment by cpitkin** [comment]
@cmcg13 Thank you for your continued support and use of the project.  

How long have you been experiencing this issue? Are you using this in Docker or as a standalone install? Can you submit a PR with the fix? 

After reviewing the code I see that line 103 is concating the oneOrLess (channels with one or fewer users) with the inactive (channels which have been inactive for less than the specified duration) array. I am curious to see what your fix looks like to see if there is a corner case I missed.

**Note:** For better tracking of issues in the future. Can you post updates as a new comment instead of editing your previous comment. That will help keep the original issue visible while also allowing for discussion and possible fixes to be seen by all.

**Comment by cmcg13** [comment]
Hey @cpitkin - I completely understand your note - I only modified the title/comment because my original comments were completely off the mark and not relevant to the actual problem. I will bear that in mind for future though - thank you!

Answers:
1. It appears it has been always happening - we didn't run a negative test until recently to ensure it was not archiving channels it should not be
2. Standalone process being run via Jenkins automated job
3. So the change I made was actually very small - I just modified it so that only the inactive array is being filtered:

`archive = _.filter(inactive, (v) => {`

I find that when we run the process, everything is going into our oneOrLess array since the logic in Lines 162-167 adds any channel to this array which has not been archived.

I added a couple of extra console.log lines just so we can debug easier, but otherwise it's still all your bot. It works really well.

**Comment by cpitkin** [comment]
This looks to be a bug introduced in v0.4.0. I will get this fixed and deploy a new version to npm in the next week.

**Comment by cpitkin** [comment]
@cmcg13 I have been testing and am unable to reproduce the issue. I have set `console.log()`several places after creating a test channel that met the <1 member criteria. Take a look at the information below and make sure I recreated your setup correctly. If you could provide your local output that would greatly help in debugging the issue. Thanks!

environment
```bash
export ARCHIVEBOT_SLACK_TOKEN=<redacted_token>
export ARCHIVEBOT_INACTIVITY_DAYS=70
export ARCHIVEBOT_NEVER_ARCHIVE=game-of-thrones,open-graph-test-channel
export ARCHIVEBOT_MEMBERS=1
```

oneOrLess array (Ln 164, post-filter)
```javascript
[ 
  { 
    id: 'CBUEW0TR8',
    name: 'testing-archive-bot',
    is_channel: true,
    created: 1532124319,
    is_archived: false,
    is_general: false,
    unlinked: 0,
    creator: 'U07889DE6',
    name_normalized: 'testing-archive-bot',
    is_shared: false,
    is_org_shared: false,
    is_member: true,
    is_private: false,
    is_mpim: false,
    topic: { value: '', creator: '', last_set: 0 },
    purpose:
     { value: 'Just to test auto archiving',
       creator: 'U07889DE6',
       last_set: 1532124320 },
    previous_names: [],
    num_members: 1
  }
]
```

Concat oneOrLess & inactive (Ln 103, pre-filter)
```javascript
[ 
  { 
    id: 'CBUEW0TR8',
    name: 'testing-archive-bot',
    is_channel: true,
    created: 1532124319,
    is_archived: false,
    is_general: false,
    unlinked: 0,
    creator: 'U07889DE6',
    name_normalized: 'testing-archive-bot',
    is_shared: false,
    is_org_shared: false,
    is_member: true,
    is_private: false,
    is_mpim: false,
    topic: { value: '', creator: '', last_set: 0 },
    purpose:
     { value: 'Just to test auto archiving',
       creator: 'U07889DE6',
       last_set: 1532124320 },
    previous_names: [],
    num_members: 1 
  }
]
```

Filtered archive array (Ln 103, post-filter)
```javascript
[ 
  { 
    id: 'CBUEW0TR8',
    name: 'testing-archive-bot',
    is_channel: true,
    created: 1532124319,
    is_archived: false,
    is_general: false,
    unlinked: 0,
    creator: 'U07889DE6',
    name_normalized: 'testing-archive-bot',
    is_shared: false,
    is_org_shared: false,
    is_member: true,
    is_private: false,
    is_mpim: false,
    topic: { value: '', creator: '', last_set: 0 },
    purpose:
      { value: 'Just to test auto archiving',
        creator: 'U07889DE6',
        last_set: 1532124320 },
    previous_names: [],
      num_members: 1 
  }
]
```

**Comment by cmcg13** [comment]
Hey Charlie - apologies for the delay, I was out on holidays for a bit. So I find that I have to set our ARCHIVEBOT_MEMBERS to greater than 1 to pull through anything from our Slack org (we don't appear to have any single member channels). Here is the full code for our Index.js class - it's now been working correctly in our Production org for 10 days:

```
#!/usr/bin/env node

const slack = require('slack');
const _ = require('underscore');
const moment = require('moment');
const chalk = require('chalk');
const program = require('commander');
const elegantStatus = require('elegant-status');
const pkgVersion = require('./package.json').version;

// Color output
const error = chalk.bold.red;
const ok = chalk.green;
const nothing = chalk.yellow;

// Get current unix time
const now = moment().unix();

// Slack API extra params
const exclude_members = true;
const exclude_archived = true;
const count = 1;
const inclusive = true;
const latest = moment.now();

// Init some arrays
let oneOrLess = [];
let inactive = [];
let archive = [];
let archivedChannelNames = [];

// Split the string
function list(val) {
  return val.split(',');
}

program
  .version(pkgVersion)
  .option('-t, --token <string>', 'Slack API bot token. You should probably use ARCHIVEBOT_SLACK_TOKEN as an ENV VAR.')
  .option('-d, --days [n]', 'Number of days of inactivity. Default: 30', parseInt)
  .option('-m, --members [n]', 'Maximum number of members in the channel. Default: 1', parseInt)
  .option('-n, --never-archive [list]', 'List of channels to never archive.', list)
  .option('-b, --basic', 'Only prints basic log messages. For use with cron or task runners. Default: false')
  .parse(process.argv);

// Sane defaults
const token = program.token || process.env.ARCHIVEBOT_SLACK_TOKEN;
const memeberCount = program.members || process.env.ARCHIVEBOT_MEMBERS || 1;
const neverArchive = program['never-archive'] || process.env.ARCHIVEBOT_NEVER_ARCHIVE || [];
const inactiveDays = program.days || process.env.ARCHIVEBOT_DAYS || 30;
const basic = program.basic || process.env.ARCHIVEBOT_BASIC || false;

// Check the token is available
if (!token) {
  console.error((error('You must provide a Slack app token. This can be done on the command line with -t, --token or set the ARCHIVEBOT_SLACK_TOKEN environment variable')));
}

function logChannels () {
  _.each(archivedChannelNames, (v) => {
    console.log((ok(`Channel ${v} has been archived. \n`)));
  })
}

function archiveChannel() {
  let done
  if (!basic) {
    done = elegantStatus(`Archiving channels -- ${archive.length}`);
  }
  if (_.size(archive) > 0) {
    let archiveIndex = -1;
    let timeout = setInterval(function () {
      ++archiveIndex
      if (archiveIndex === archive.length) {
        if (!basic) {
          done(true);
        }
        clearInterval(timeout);
      } else {
        if (!basic) {
          done.updateText(`Archiving channels -- ${archiveIndex+1}/${archive.length}`)
        }
        let channel = archive[archiveIndex].id
        slack.channels.archive({ token, channel }, (err) => {
          if (err) {
            if (!basic) {
              done.updateText(error(`Error archiving channel ${channel.name} with error: ${err}`));
              done(false);
            } else {
              console.error(error(`Error archiving channel ${channel.name} with error: ${err}`));
            }
          }
          archivedChannelNames.push(archive[archiveIndex].name)
        });
      }
    }, 2000)
    logChannels();
  } else {
    console.log((nothing('No channels met the archive criteria. \n')));
  }
}

function filterChannels() {
  console.log(`Inactive Array: ${_.size(inactive)}`);
  archive = _.filter(inactive, (v) => {
    if (!_.contains(neverArchive, v.name)) {
      return v;
    }
  });
  console.log(`Archive Array: ${_.size(archive)}`);
  archiveChannel();
}

function getHistory(channels) {
  let done
  if (!basic) {
    done = elegantStatus(`Fetching channel history for ${channels.length} channel(s)`);
  }
  let channelIndex = -1;
  let timeout = setInterval(function () {
    ++channelIndex
    if (channelIndex === channels.length) {
      if (!basic) {
        done(true);
      }
      filterChannels();
      clearInterval(timeout)
    } else {
      if (!basic) {
        done.updateText(`Fetching channel history -- ${channelIndex+1}/${channels.length}`)
      }
      let channel = channels[channelIndex].id
      slack.channels.history({ token, channel, count, inclusive, latest }, (err, data) => {
        if (err) {
          if (!basic) {
            done.updateText(error(`Error fetching channel ${channel.name} with ${err}`));
            done(false);
          } else {
            console.error(error(`Error fetching channel ${channel.name} with ${err}`));
          }
        }
        console.log(`Inactive Days: ${inactiveDays}`);
        let f = moment.duration(now - data.messages[0].ts, 's').asDays();
        console.log(`Moment Calculation: ${f}`);
        console.log(`Channel Name: ${channels[channelIndex].name}`);
        if (f > inactiveDays) {
          data.id = channels[channelIndex].id
          data.name = channels[channelIndex].name
          inactive.push(data)
        }
      })
    }
  }, 2000);
}

slack.channels.list({ token, exclude_archived, exclude_members }, (err, data) => {
  let done
  if (!basic) {
    done = elegantStatus('Fetching Slack channels')
  }
  if (err) {
    if (!basic) {
      done.updateText(error(`Error fetching channel list with error: ${err}`));
      done(false)
    } else {
      console.error(error(`Error fetching channel list with error: ${err}`));
    }
  }
  oneOrLess = _.filter(data.channels, (v) => {
    if (v.num_members <= memeberCount) {
      return v;
    } else if (v.is_archived !== false) {
      return v;
    }
  });
  if (!basic) {
    done(true)
  }
  getHistory(oneOrLess);
});
```

Let me know if you need anything else but the real crux of the issue was that it correctly pulled all the channels, completed the History check logic and then it was just adding everything from the original array back in and disregarding the History check logic - once I corrected that it performed as expected.

**Comment by cpitkin** [comment]
@cmcg13 I just tagged a pre-release v0.4.1-pre. If you could test it out that would be great. Thanks!

**Comment by cmcg13** [comment]
Hey Charlie - this new version looks spot on to me. Thanks!

**Comment by cpitkin** [comment]
Fixed in v0.4.1
